keps/sig-node/compute-device-assignment.md

```diff
@@ -0,0 +1,146 @@
+* As a _Device Vendor_, I manufacture devices and I have deep domain expertise in how to run and monitor them. Because I maintain my own Device Plugin implementation, as well as Device Monitoring Agent, I can provide consumers of my devices an easy way to consume and monitor my devices without requiring open-source contributions. The Device Monitoring Agent doesn't have any dependencies on the Device Plugin, so I can decouple monitoring from device lifecycle management. My Device Monitoring Agent works by periodically querying the `/devices/<ResourceName>` endpoint to discover which devices are being used, and to get the container/pod metadata associated with the metrics:
+
+![device monitoring architecture](https://user-images.githubusercontent.com/3262098/43926483-44331496-9bdf-11e8-82a0-14b47583b103.png)
+
+
+## Changes
+
+Add a v1alpha1 Kubelet GRPC service, at `/var/lib/kubelet/pod-resources/kubelet.sock`, which returns information about the kubelet's assignment of devices to containers. It obtains this information from the internal state of the kubelet's Device Manager. The GRPC Service returns a single PodResourcesResponse, which is shown in proto below:
+```protobuf
+// PodResources is a service provided by the kubelet that provides information about the
+// node resources consumed by pods and containers on the node
+service PodResources {
+    rpc List(ListPodResourcesRequest) returns (ListPodResourcesResponse) {}
+}
+
+// ListPodResourcesRequest is the request made to the PodResources service
+message ListPodResourcesRequest {}
+
+// ListPodResourcesResponse is the response returned by List function
+message ListPodResourcesResponse {
+    repeated PodResources pod_resources = 1;
+}
+
+// PodResources contains information about the node resources assigned to a pod
+message PodResources {
+    string name = 1;
+    string namespace = 2;
+    repeated ContainerResources containers = 3;
+}
+
+// ContainerResources contains information about the resources assigned to a container
+message ContainerResources {
+    string name = 1;
+    repeated ContainerDevices devices = 2;
+}
+
+// ContainerDevices contains information about the devices assigned to a container
+message ContainerDevices {
+    string resource_name = 1;
+    repeated string device_ids = 2;
+}
+```
+
+### Potential Future Improvements
+
+* Add `ListAndWatch()` function to the GRPC endpoint so monitoring agents don't need to poll.
+* Add identifiers for other resources used by pods to the `PodResources` message.
+  * For example, persistent volume location on disk
+
+## Alternatives Considered
+
+### Add a `ContainerConfig` function to the CRI, which returns the [CreateContainerRequest](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L734) used to create the container.
```

Comment:
How are devices being passed down from the kubelet to the runtimes today if CRi doesn't support specifying devices explicitly? If kubelet can expose to monitoring agents the config blob that it passes down to runtimes, that might be another alternate endpoint you can expose instead of manually curating a separate endpoint.

Reply:
CRI does support specifying devices directly: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L670.  But the kubelet doesn't keep that information around.  We would have to persist (checkpoint) this in order to support such an endpoint.  While it is nice not to need to curate another API, I personally prefer a more limited API that does not require checkpointing additional information.

Reply:
If the kubelet isn't checkpointing, then it presumably has some meant to determinisitically regenerate this data?

Reply:
True, we could regenerate this data.

Reply:
Ok. So in that case the rationale for exposing an alternate endpoint would be:
1. Restrict what data plugins can access - the policy around this is unclear - what information is considered to be sensitive?
2. Deluge of information on the plugin side - is that really a problem given you are embracing protobufs which should reduce the memory requirements on the plugin side?

Reply:
I updated this section to take this into account.  I still would rather define a new API, since the `ContainerConfig` doesn't have much of what we need outside of device information.

Reply:
I trust your instinct. Thanks for being precise about the design considerations.

Reply:
1. It isn't about information, it is about the ability to create containers.  Anyone with access to the CRI socket can create containers, which is effectively root on the node.
2. Isn't an issue.
3. More of a minor issue.